File: app/src/main/kotlin/dev/trindade/robokide/ui/syntax/SimpleHighlighter.java
package dev.trindade.robokide.ui.syntax;

import android.text.Editable;
import android.text.Spannable;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.TextWatcher;
import android.text.style.CharacterStyle;
import android.text.style.ForegroundColorSpan;
import android.widget.EditText;
import android.widget.TextView;

import java.util.List;
import java.util.regex.Matcher;

public class SimpleHighlighter {
    
    private final List<SyntaxScheme> syntaxList;
    private final EditText mEditor;
    private final TextView mTextView;
    
    private String type;

    public SimpleHighlighter(EditText editor, String syntaxType) {
        this.mEditor = editor;
        this.mTextView = null;
        this.syntaxList = getSyntaxList(syntaxType);
        init();
    }

    public SimpleHighlighter(TextView textView, String syntaxType) {
        this.mTextView = textView;
        this.mEditor = null;
        this.syntaxList = getSyntaxList(syntaxType);
        init();
    }

    


    private List<SyntaxScheme> getSyntaxList(String syntaxType) {
        switch (syntaxType.toLowerCase()) {
            case "xml":
                return SyntaxScheme.XML();
            case "java":
                return SyntaxScheme.JAVA();
            default:
                throw new IllegalArgumentException("Unsupported syntax type");
        }
    }


    private void init() {
        if (mEditor != null) {
            mEditor.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                }

                @Override
                public void afterTextChanged(Editable s) {
                    removeSpans(s, ForegroundColorSpan.class);
                    createHighlightSpans(syntaxList, s);
                }
            });
            removeSpans(mEditor.getText(), ForegroundColorSpan.class);
            createHighlightSpans(syntaxList, mEditor.getText());
        } else if (mTextView != null) {
            CharSequence text = mTextView.getText();
            if (text instanceof Editable) {
                Editable editable = (Editable) text;
                removeSpans(editable, ForegroundColorSpan.class);
                createHighlightSpans(syntaxList, editable);
            } else {
                SpannableStringBuilder builder = new SpannableStringBuilder(text);
                removeSpans(builder, ForegroundColorSpan.class);
                createHighlightSpans(syntaxList, builder);
                mTextView.setText(builder);
            }
        }
    }

    private void createHighlightSpans(List<SyntaxScheme> syntaxList, Editable editable) {
        for (SyntaxScheme scheme : syntaxList) {
            for (Matcher m = scheme.pattern.matcher(editable); m.find(); ) {
                if (scheme == scheme.getPrimarySyntax()) {
                    editable.setSpan(new ForegroundColorSpan(scheme.color), m.start(), m.end() - 1, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                } else {
                    editable.setSpan(new ForegroundColorSpan(scheme.color), m.start(), m.end(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                }
            }
        }
    }

    private void removeSpans(Editable editable, Class<? extends CharacterStyle> type) {
        CharacterStyle[] spans = editable.getSpans(0, editable.length(), type);
        for (CharacterStyle span : spans) {
            editable.removeSpan(span);
        }
    }

    private void removeSpans(Spannable spannable, Class<? extends CharacterStyle> type) {
        CharacterStyle[] spans = spannable.getSpans(0, spannable.length(), type);
        for (CharacterStyle span : spans) {
            spannable.removeSpan(span);
        }
    }
}

File: app/src/main/kotlin/dev/trindade/robokide/ui/syntax/SimpleHighlighter.kt
package dev.trindade.robokide.ui.syntax

import android.text.*
import android.text.style.CharacterStyle
import android.text.style.ForegroundColorSpan
import android.widget.EditText
import android.widget.TextView

class SimpleHighlighter(private val mEditor: EditText?, syntaxType: String) {

    private val syntaxList: List<SyntaxScheme>
    private val mTextView: TextView? = null

    init {
        this.syntaxList = getSyntaxList(syntaxType)
        init()
    }

    constructor(textView: TextView, syntaxType: String) : this(null, syntaxType) {
        mTextView = textView
        init()
    }

    private fun getSyntaxList(syntaxType: String): List<SyntaxScheme> {
        return when (syntaxType.toLowerCase()) {
            "xml" -> SyntaxScheme.XML()
            "java" -> SyntaxScheme.JAVA()
            else -> throw IllegalArgumentException("Unsupported syntax type")
        }
    }

    private fun init() {
        mEditor?.let { editor ->
            editor.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
                override fun afterTextChanged(s: Editable) {
                    removeSpans(s, ForegroundColorSpan::class.java)
                    createHighlightSpans(syntaxList, s)
                }
            })
            removeSpans(editor.text, ForegroundColorSpan::class.java)
            createHighlightSpans(syntaxList, editor.text)
        } ?: mTextView?.let { textView ->
            val text = textView.text
            if (text is Editable) {
                removeSpans(text, ForegroundColorSpan::class.java)
                createHighlightSpans(syntaxList, text)
            } else {
                val builder = SpannableStringBuilder(text)
                removeSpans(builder, ForegroundColorSpan::class.java)
                createHighlightSpans(syntaxList, builder)
                textView.text = builder
            }
        }
    }

    private fun createHighlightSpans(syntaxList: List<SyntaxScheme>, editable: Editable) {
        for (scheme in syntaxList) {
            val matcher = scheme.pattern.matcher(editable)
            while (matcher.find()) {
                if (scheme == scheme.primarySyntax) {
                    editable.setSpan(ForegroundColorSpan(scheme.color), matcher.start(), matcher.end() - 1, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                } else {
                    editable.setSpan(ForegroundColorSpan(scheme.color), matcher.start(), matcher.end(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                }
            }
        }
    }

    private fun removeSpans(editable: Editable, type: Class<out CharacterStyle>) {
        val spans = editable.getSpans(0, editable.length, type)
        for (span in spans) {
            editable.removeSpan(span)
        }
    }

    private fun removeSpans(spannable: Spannable, type: Class<out CharacterStyle>) {
        val spans = spannable.getSpans(0, spannable.length, type)
        for (span in spans) {
            spannable.removeSpan(span)
        }
    }
}